### PR TITLE
sync poet messages

### PR DESCRIPTION
**Pull request by noamnelke**

### CODE REVIEW

File: sync/syncer.go [hunk 1]
```diff
@@ -375,6 +394,21 @@ func (s *Syncer) ATXs(mb *types.Block) (atxs []*types.ActivationTx, associated *
 	return atxs, associated, nil
 }
 
+func (s *Syncer) EnsurePoetProofAvailableAndValid(poetProofRef []byte) error {
```
Comment (almogdepaz) [inline sync/syncer.go]: please shorten the method names
PoetProofAvailable could be something like "Exist" you don't have to include "poet" in the name
it is implied by the var name  poetDb :) 
"EnsurePoetProofAvailableAndValid" should be something like FetchPoetProof or SyncPoetProof

File: activation/poetdb.go [hunk 2]
```diff
@@ -29,18 +28,28 @@ func NewPoetDb(store database.Database, log log.Log) *PoetDb {
 	return &PoetDb{store: store, poetProofRefSubscriptions: make(map[poetProofKey][]chan []byte), log: log}
 }
 
-type processingError string
+func (db *PoetDb) PoetProofAvailable(proofRef []byte) bool {
+	_, err := db.store.Get(proofRef)
+	return err == nil
+}
+
+func (db *PoetDb) ValidateAndStorePoetProof(proofMessage types.PoetProofMessage) error {
```
Comment (almogdepaz) [inline activation/poetdb.go]: ValidateAndStore

File: activation/poetdb.go [hunk 3]
```diff
@@ -127,16 +136,20 @@ func (db *PoetDb) publishPoetProofRef(key poetProofKey, poetProofRef []byte) {
 	delete(db.poetProofRefSubscriptions, key)
 }
 
-func (db *PoetDb) GetMembershipByPoetProofRef(poetRef []byte) (map[common.Hash]bool, error) {
-	poetProofBytes, err := db.store.Get(poetRef)
+func (db *PoetDb) GetPoetProofMessage(proofRef []byte) ([]byte, error) {
```
Comment (almogdepaz) [inline activation/poetdb.go]: you can shorten this as well to "Get" poetProof is implied

File: activation/poetdb.go [hunk 4]
```diff
@@ -127,16 +136,20 @@ func (db *PoetDb) publishPoetProofRef(key poetProofKey, poetProofRef []byte) {
 	delete(db.poetProofRefSubscriptions, key)
 }
```
Comment (almogdepaz) [inline activation/poetdb.go]: you can shorten this to Get the poet part is implied
also you should use this method inside GetMembershipByPoetProofRef
i would change "GetMembershipByPoetProofRef" to GetAsMessage or something that implies the conversion from bytes

File: activation/poetdb.go (hunk 3 above)
Comment (almogdepaz) [inline activation/poetdb.go]: ou can shorten this to Get the poet part is implied
also you should use this method inside GetMembershipByPoetProofRef
i would change "GetMembershipByPoetProofRef" to GetAsMessage or something that implies the conversion from bytes

File: activation/poetdb.go [hunk 5]
```diff
@@ -29,18 +28,28 @@ func NewPoetDb(store database.Database, log log.Log) *PoetDb {
 	return &PoetDb{store: store, poetProofRefSubscriptions: make(map[poetProofKey][]chan []byte), log: log}
 }
 
-type processingError string
+func (db *PoetDb) HasProof(proofRef []byte) bool {
+	_, err := db.GetProofMessage(proofRef)
+	return err == nil
+}
+
+func (db *PoetDb) ValidateAndStore(proofMessage types.PoetProofMessage) error {
```
Comment (zalmen) [inline activation/poetdb.go]: Shouldn't the poet proof be passed by reference?

File: activation/poetdb.go [hunk 6]
```diff
@@ -52,48 +61,48 @@ func (db *PoetDb) ValidatePoetProof(proof types.PoetProof, poetId [types.PoetIdL
 	return nil
 }
 
-func (db *PoetDb) storePoetProof(proof types.PoetProof, poetId [types.PoetIdLength]byte, roundId uint64,
-	signature []byte) error {
-
-	poetProof, err := types.InterfaceToBytes(&proof)
+func (db *PoetDb) storeProof(proofMessage types.PoetProofMessage) error {
```
Comment (zalmen) [inline activation/poetdb.go]: Shouldn't the poet proof be passed by reference?

File: activation/poetdb.go [hunk 7]
```diff
@@ -52,48 +61,48 @@ func (db *PoetDb) ValidatePoetProof(proof types.PoetProof, poetId [types.PoetIdL
...
+	poetProofBytes, err := types.InterfaceToBytes(&proofMessage.PoetProof)
 	if err != nil {
 		return fmt.Errorf("failed to marshal poet proof for poetId %x round %d: %v",
-			poetId, roundId, err)
+			proofMessage.PoetId, proofMessage.RoundId, err)
 	}
 
-	ref := sha256.Sum256(poetProof)
+	ref := sha256.Sum256(poetProofBytes)
```
Comment (zalmen) [inline activation/poetdb.go]: it shouldn't work this way, we should have a single place that calculates references. Don't change it now but we should have a technical debt issue to handle it

File: miner/block_builder.go [hunk 8]
```diff
@@ -35,6 +35,10 @@ type AtxValidator interface {
 	SyntacticallyValidateAtx(atx *types.ActivationTx) error
 }
 
+type Syncer interface {
+	SyncPoetProof(poetProofRef []byte) error
```
Comment (zalmen) [inline miner/block_builder.go]: Rename to fetch?

File: miner/block_builder.go [hunk 9]
```diff
@@ -251,6 +258,13 @@ func (t *BlockBuilder) listenForAtx() {
 					t.Panic("nil nipst in gossip")
 					break
 				}
+
+				if err := t.syncer.SyncPoetProof(atx.GetPoetProofRef()); err != nil {
```
Comment (zalmen) [inline miner/block_builder.go]: it is a long blocking operation, it should be called in a go routine

File: sync/syncer.go [hunk 10]
```diff
@@ -351,6 +361,15 @@ func (s *Syncer) ATXs(mb *types.Block) (atxs []*types.ActivationTx, associated *
...
 			ntxs := out.([]types.ActivationTx)
 			for _, atx := range ntxs {
+				err := s.SyncPoetProof(atx.GetPoetProofRef())
+				if err != nil {
+					if types.IsProcessingError(err) {
+						s.Error("error while validating PoET proof: %v", err)
+					} else {
+						s.Warning("received ATX with invalid or missing PoET proof: %v", err)
+					}
+					return nil, nil, fmt.Errorf("a referenced PoET proof is missing or invalid: %v", err)
```
Comment (zalmen) [inline sync/syncer.go:371]: IMO we shouldn't halt on the first failure. Instead, we should fetch all ATXs, store those who failed and print all of them. It will ease the debugging

**Review by zalmen — CHANGES_REQUESTED** [review]

File: sync/syncer.go (hunk 10 above)
Comment (noamnelke) [reply sync/syncer.go:371]: I'm not sure it will ease debugging. I think it could cause weird situations that are harder to debug, e.g.:

We try to fetch the PoET proof of an ATX and fail. We keep that ATX ID for later and move on. We try to fetch the PoET proof for the next ATX, which happens to be the same one, essentially retrying. One of our neighbors now got this PoET proof and sends it to us. We then later print the list of ATXs that failed and it now doesn't make sense, since we see that we previously stored the PoET proof (after the retry).

Generally, I find that the simpler the logic, the easier the debugging.

File: activation/poetdb.go [hunk 11]
```diff
@@ -29,18 +28,28 @@ func NewPoetDb(store database.Database, log log.Log) *PoetDb {
...
 }
 
-func (db *PoetDb) ValidatePoetProof(proof types.PoetProof, poetId [types.PoetIdLength]byte, roundId uint64,
+func (db *PoetDb) Validate(proof types.PoetProof, poetId [types.PoetIdLength]byte, roundId uint64,
 	signature []byte) error {
 
 	root, err := calcRoot(proof.Members)
 	if err != nil {
-		return processingError(fmt.Sprintf("failed to calculate membership root for poetId %x round %d: %v",
+		return types.ProcessingError(fmt.Sprintf("failed to calculate membership root for poetId %x round %d: %v",
```
Comment (antonlerner) [inline activation/poetdb.go:52]: shorten poet id

File: activation/poetdb.go [hunk 12]
```diff
@@ -127,16 +136,20 @@ func (db *PoetDb) publishPoetProofRef(key poetProofKey, poetProofRef []byte) {
...
-	poetProofBytes, err := db.store.Get(poetRef)
+func (db *PoetDb) GetProofMessage(proofRef []byte) ([]byte, error) {
+	return db.store.Get(proofRef)
+}
+
+func (db *PoetDb) GetMembershipMap(proofRef []byte) (map[common.Hash]bool, error) {
+	proofMessageBytes, err := db.GetProofMessage(proofRef)
 	if err != nil {
-		return nil, fmt.Errorf("could not fetch poet proof for ref %x: %v", poetRef, err)
+		return nil, fmt.Errorf("could not fetch poet proof for ref %x: %v", proofRef, err)
```
Comment (antonlerner) [inline activation/poetdb.go]: shorten ref id

File: activation/poetdb.go [hunk 13]
```diff
@@ -127,16 +136,20 @@ func (db *PoetDb) publishPoetProofRef(key poetProofKey, poetProofRef []byte) {
 	delete(db.poetProofRefSubscriptions, key)
 }
 
-func (db *PoetDb) GetMembershipByPoetProofRef(poetRef []byte) (map[common.Hash]bool, error) {
-	poetProofBytes, err := db.store.Get(poetRef)
+func (db *PoetDb) GetProofMessage(proofRef []byte) ([]byte, error) {
+	return db.store.Get(proofRef)
```
Comment (antonlerner) [inline activation/poetdb.go:137]: could this be called in a go routine? shoould we lock here if so?
Have you run this code with --race?

Comment (noamnelke) [reply activation/poetdb.go:137]: while `--race` breaks many of our unit tests (23 of them), it doesn't look like any of the failures are related to the changes in this PR.

**Review by zalmen — APPROVED** [review]
